### PR TITLE
updated interactor modeling workflow tutorial, lasso_modeling source code and test suite

### PR DESCRIPTION
**Pull request by ejiawustl**
This addresses the following

1) My current process of LassoCV followed by backwards selection needs to be implemented
2) The three functions at the bottom that perform the OLS CV avg(r2) and interactor analysis need to be read and confirmed for accuracy, and then the output needs to be adjusted based on the analysis you do.
3) Those three functions, and any that are necessary for step 1, need to then be moved into source code.
4) Tests added for all functions moved into the source code

For the interactor modeling workflow tutorial notebook, you will notice a big difference which is that I have added an entire separate workflow after the end of the first one. Additionally, I made some small changes to Workflow 1 (namely when testing interactor variants to only test subbing in the main effect) that better align with the final pipeline given in the Yeast To-Dos doc. Workflow 2 comes with complete description and follows my original pipeline of LassoCV, intersecting terms, then performing backwards OLS selection. All helper methods and functions have been added to the source code with proper docstring and typehinting,

### CODE REVIEW

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 1]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
+        prev_set_size = curr_set_size
+        curr_set_size = len(curr_feature_set)
+
+    # Reinitialize variables for the second loop
+    prev_set_size = 0  # Reset for Step 3.2
+    curr_set_size = len(curr_feature_set)
+
+    # Step 3.2: Iteratively filter features using the top 10% dataset
+    # this time we use a less stringent p-value threshold since there is less data
+    while curr_set_size != prev_set_size and curr_set_size != 0:
```
Comment (cmatKhan) [inline yeastdnnexplorer/ml_models/lasso_modeling.py:921]: This appears to be repeated code from line 906. Unless there is a good argument against, write this in a way that you can pass the y,X and a threshold in in the arguments

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 2]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
+def select_significant_features(
+    perturbed_tf: str, feature_set: set, data: pd.DataFrame, p_value_threshold: float
+) -> list:
+    """
+    This is a helper method for backwards_OLS_feature_selection which select significant
+    features from a given set of predictors by creating an OLS model and removing
+    features whose p-values are above a given threshold. This performs a single
+    iteration of filtering, and returns the set of interactors that survive after.
+
+    Note: because the data input into this method was created by get_full_data(), which
```
Comment (cmatKhan) [inline yeastdnnexplorer/ml_models/lasso_modeling.py:984]: Will the features always be a subset of the original `X`, which you would have already generated if you think of the workflow which processes a single TF from start to finish? I believe yes. If that is the case, that  `X` will already exist in your environment, and you can do this bu selecting feature columns rather than regenerating a new dataframe. There is an example of something like this  in `get_interactor_importance`. That would eliminate this additional code surrounding building a formula

**Review by cmatKhan — CHANGES_REQUESTED** [review]
1. There are functions in here that do not belong in source code -- please do consider what is more useful as part of the API, and what code should be placed in the tutorial to show the user how to use the API.

2. It seems like there is a function that executes the same code on both "all data" and the "top10%". Remember that when you're working in the source code, if you run into a pattern like that, to think critically about whether the code could be simplified so that it works on both, most likely by adjusting what is input.

3. Please consider output in light of analysis and adjust accordingly.


Finally, `lasso_modeling.py` is no longer an accurate description of the scope of work in this module. Please propose a solution as a response to this comment

**Comment by ejiawustl** [comment]
Re: Finally, lasso_modeling.py is no longer an accurate description of the scope of work in this module. Please propose a solution as a response to this comment.

Given that the tutorial is called interactor_modeling_workflow, would an appropriate file name be interactor_modeling.py or interactor_modeling_workflows.py?

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 3]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
     return fig, significant_coefs_dict
+
+
+# TODO: must have option to add max_lrb
+
+# TODO: the top10% models likely should not have teh same number of classes as the
+# all data, possibly not stratified at all
+
+
+def get_significant_predictors(
```
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py]: removed from source code and test suite in latest pull

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 4]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
+        model = estimator_local.fit(X_train, y_train)
+
+        # Calculate R-squared and append to r2_scores
+        r2_scores.append(r2_score(y_test, model.predict(X_test)))
+
+    return np.mean(r2_scores)
+
+
+def try_interactor_variants(
+    intersect_coefficients, interactor, **kwargs
```
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py]: Added typing in latest pull

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 5]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
+    if X is None:
+        raise ValueError("X must be passed as a keyword argument")
+    stratification_classes = kwargs.get("stratification_classes")
+    if stratification_classes is None:
+        raise ValueError("stratification_classes must be passed as a keyword argument")
+
+    main_effect = interactor.split(":")[1]
+    # interactor_formula_variants = [main_effect, [main_effect, interactor]]
+
+    # NOTE: may need to change this back to the original line above: I just noticed in
```
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py]: Removed the comments and addressed in latest pull

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 6]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
+    for variant in interactor_formula_variants:
+        # replace the interactor term in the formula with the variant
+        variant_predictors = (
+            [term for term in intersect_coefficients if term != interactor] + [variant]
+            if isinstance(variant, str)
+            else variant
+        )
+        # conduct the stratified CV r-squared calculation with the formula variant
+        input_model_avg_rsquared = stratified_cv_r2(
+            y, X.loc[:, variant_predictors], stratification_classes
```
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py:631]: Added the intercept into stratified_cv_r2 if it isn't there -- I checked this with respect to the entire workflow and now it should handle intercept terms appropriately. Intercept terms are never included as significant features

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 7]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
+
+        # append the results to the output list
+        output.append(
+            {
+                "interactor": interactor,
+                "variant": variant,
+                "avg_r2": input_model_avg_rsquared,
+            }
+        )
+    return output
```
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py:642]: I am going to keep the output the same for now unless there are objections. I have a draft of a for loop I would use across all tfs to gather their avg r-squared values from the final models. Additionally, in the tutorial it currently shows how do pull the relevant bits of information from the output for the analysis we are interested in. I can change this if needed after further discussion, but I think it is *alright* for now

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 8]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
+                stratification_classes=stratification_classes,
+            )
+
+            # compare the avg_r2 values of the two variants to input_model_avg_rsquared
+            variant_dict = max(interactor_variant_results, key=lambda x: x["avg_r2"])
+
+            if variant_dict["avg_r2"] > input_model_avg_rsquared:
+                interactor_results.append(variant_dict)
+
+    return input_model_avg_rsquared, interactor_results
```
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py:694]: Same comment as above.

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 9]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
+
+
+def get_non_zero_predictors(
```
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py]: Added this to tutorial and removed from source code and test suite.

File: yeastdnnexplorer/ml_models/lasso_modeling.py [hunk 10]
```diff
@@ -530,3 +533,500 @@ def examine_bootstrap_coefficients(
...
+                perturbed_tf, curr_feature_set, top_10_data, p_value_threshold=0.01
+            )
+        )
+        prev_set_size = curr_set_size
+        curr_set_size = len(curr_feature_set)
+
+    return curr_feature_set
+
+
+def get_full_data(
```
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py]: I removed this function entirely -- you were right, I found a way to essentially perform this process more effectively in the backward_OLS_feature_selection() function which was where it was called originally

File: yeastdnnexplorer/ml_models/lasso_modeling.py (hunk 1 above)
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py]: I have re-factored and reparameterized this code following our discussion. It now takes in a list of data quantile thresholds and p-value thresholds and performs OLS backward feature selection using these user inputs instead of hardcoding anything in.

File: yeastdnnexplorer/ml_models/lasso_modeling.py (hunk 2 above)
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py]: Have re-written this helper function's original function to pass in this data which was generated in that function. That should address this.

File: yeastdnnexplorer/ml_models/lasso_modeling.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (cmatKhan) [reply yeastdnnexplorer/ml_models/lasso_modeling.py:631]: I'm not so concerned about them being included as i am with the intercept being excluded. In the current LassoCV method, we are not even considering whether the intercept is significant or not.

I'm of the opinion that it should be included no matter what. However, it would be good to know how this is being handled in your analysis notebook (not the source code tutorial that you have been working on for this code).

File: yeastdnnexplorer/ml_models/lasso_modeling.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (cmatKhan) [reply yeastdnnexplorer/ml_models/lasso_modeling.py:642]: If you need coefficients or some other metric that could be extracted at this point, it would be best to do so rather than running the same model again somewhere else, if that is what you mean. I need to look at the notebook still, though

**Review by cmatKhan — CHANGES_REQUESTED** [review]

**Comment by cmatKhan** [comment]
> interactor_modeling

I like interactor_modeling.py -- I'll make that change when I do the entry point script

File: yeastdnnexplorer/ml_models/lasso_modeling.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py:694]: I thought about it, and I think it should be fine -- any necessary coefficients or pvalues could be generated by creating the statsmodel multivariate model: currently, in the call to get the average R-squared from cross validation from the model, it doesn't output anything it addition to it. I'm thinking it will be fine to keep the outputs how they are -- if we needed to get any additional information, it is able to be generated from creating a single linear model - we wouldn't be re-running anything per se because the current function to perform stratified CV only outputs the r-squared; I think that is okay for now. 

TLDR: if we needed to do additional analysis, we could take the features in our output and make linear models to get these metrics (pvals, coefs, etc.) because the existing methods do not do this. I think it is okay to leave everything as is but we can discuss more if you have a different idea.

File: yeastdnnexplorer/ml_models/lasso_modeling.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ejiawustl) [reply yeastdnnexplorer/ml_models/lasso_modeling.py:642]: See comment below -- also let me know your thoughts after looking through the notebook
